org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/text/correction/QuickAssistProcessor.java

```diff
@@ -121,6 +224,42 @@ private static int getIndex(int offset, List<Statement> statements) {
 		return statements.size();
 	}
 
+	private static boolean getExtractMethodProposal(IInvocationContext context, ASTNode coveringNode, boolean problemsAtLocation, Collection<CUCorrectionProposal> proposals) throws CoreException {
+		if (!(coveringNode instanceof Expression) && !(coveringNode instanceof Statement) && !(coveringNode instanceof Block)) {
+			return false;
+		}
+		if (coveringNode instanceof Block) {
+			List<Statement> statements = ((Block) coveringNode).statements();
+			int startIndex = getIndex(context.getSelectionOffset(), statements);
+			if (startIndex == -1) {
+				return false;
+			}
+			int endIndex = getIndex(context.getSelectionOffset() + context.getSelectionLength(), statements);
+			if (endIndex == -1 || endIndex <= startIndex) {
+				return false;
+			}
+		}
+
+		if (proposals == null) {
+			return true;
+		}
+
+		final ICompilationUnit cu = context.getCompilationUnit();
+		final ExtractMethodRefactoring extractMethodRefactoring = new ExtractMethodRefactoring(context.getASTRoot(), context.getSelectionOffset(), context.getSelectionLength());
+		extractMethodRefactoring.setMethodName("extracted"); //$NON-NLS-1$

```

Comment:
the method name should be incremented if there's already an `extracted()` method, with the same parameters

Reply:
I haven't checked how the TS server does it, but the rename command is invoked right after the refactoring for typescript: 
https://code.visualstudio.com/updates/v1_16#_javascript-and-typescript-refactorings

Reply:
Yes, the typescript/javascript will trigger after rename command after the rename. I'll see if we need implement in this PR or another standalone PR. 

As for duplicate method name,  I think if rename/selection can be triggered after the refactor, it is not necessary to introduce the mechanism to avoid the duplication, same behavior as eclipse.